File: fsharp/FsSearch/SearchOption.fs
namespace FsSearch

open System
open System.Collections.Generic
open System.IO
open System.Text
open System.Text.RegularExpressions

type SearchOption(shortArg : string, longArg : string, description : string) =
    let _shortArg = shortArg
    let _longArg = longArg
    let _description = description

    // read-only member properties
    member this.ShortArg = _shortArg
    member this.LongArg = _longArg
    member this.Description = _description
    member this.SortArg = 
        if _shortArg <> null && _shortArg <> "" then _shortArg
        else _longArg


//type SearchArgOption(shortArg : string, longArg : string, description : string, action : Action<string, SearchSettings>) =
type SearchArgOption(shortArg : string, longArg : string, description : string, action : Action<string, SearchSettings>) =
    inherit SearchOption(shortArg, longArg, description)
    let _action = action
    member this.Action = _action

type SearchFlagOption(shortArg : string, longArg : string, description : string, action : Func<SearchSettings, bool>) =
    inherit SearchOption(shortArg, longArg, description)
    let _action = action
    member this.Action = _action
    ;;

File: csharp/CsSearch/CsSearch/SearchOption.cs
using System;

namespace CsSearch
{
	public class SearchOption
	{
		public string ShortArg { get; private set; }
		public string LongArg { get; private set; }
		public string SortArg
		{
			get
			{
				var longArg = LongArg.Replace("in-", "ina");
				if (!string.IsNullOrWhiteSpace(ShortArg))
					return ShortArg.ToLower() + "a" + longArg;
				return longArg;
			}
		}
		public string Description { get; private set; }

		public SearchOption(string shortArg, string longArg, string description)
		{
			ShortArg = shortArg;
			LongArg = longArg;
			Description = description;
		}
	}

	class SearchArgOption : SearchOption
	{
		public Action<string, SearchSettings> Action { get; private set; }
		public SearchArgOption(string shortArg, string longArg, Action<string, SearchSettings> action, string description) :
			base(shortArg, longArg, description)
		{
			Action = action;
		}
	}

	class SearchFlagOption : SearchOption
	{
		public Action<bool, SearchSettings> Action { get; private set; }
		public SearchFlagOption(string shortArg, string longArg, Action<bool, SearchSettings> action, string description) :
			base(shortArg, longArg, description)
		{
			Action = action;
		}
	}
}
